src/main/java/org/aksw/simba/squirrel/data/uri/filter/RDBKnownUriFilter.java

```diff
@@ -151,16 +165,59 @@ public void add(CrawleableUri uri, long lastCrawlTimestamp, long nextCrawlTimest
             r.db(DATABASE_NAME).table(TABLE_NAME).filter(doc -> doc.getField(COLUMN_URI).eq(uri.getUri().toString())).update(r.hashMap(COLUMN_TIMESTAMP_LAST_CRAWL, lastCrawlTimestamp));
+            r.db(DATABASE_NAME).table(TABLE_NAME).filter(doc -> doc.getField(COLUMN_URI).eq(uri.getUri().toString())).update(r.hashMap(COLUMN_HASH_VALUE, DUMMY_HASH_VALUE));
             r.db(DATABASE_NAME).table(TABLE_NAME).filter(doc -> doc.getField(COLUMN_URI).eq(uri.getUri().toString())).update(r.hashMap((COLUMN_TIMESTAMP_NEXT_CRAWL), nextCrawlTimestamp)).run(connector.connection);
         } else {
             r.db(DATABASE_NAME)
                 .table(TABLE_NAME)
-                .insert(convertURITimestampToRDB(uri, lastCrawlTimestamp, nextCrawlTimestamp, false))
+                .insert(convertURITimestampToRDB(uri, lastCrawlTimestamp, nextCrawlTimestamp, false, DUMMY_HASH_VALUE))
                 .run(connector.connection);
         }
         LOGGER.debug("Adding URI {} to the known uri filter list", uri.toString());
     }
 
+    @Override
+    public void addHashValuesForUris(List<CrawleableUri> uris) {
+        for (CrawleableUri uri : uris) {
+            r.db(DATABASE_NAME).table(TABLE_NAME).filter(doc -> doc.getField(COLUMN_URI).eq(uri.getUri().toString())).
+                update(r.hashMap(COLUMN_HASH_VALUE, uri.getHashValue().encodeToString())).run(connector.connection);
+        }
+    }
+
+    @Override
+    public List<CrawleableUri> getAllUris() {
+        Cursor<HashMap> cursor = r.db(DATABASE_NAME).table(TABLE_NAME).run(connector.connection);
+        List<CrawleableUri> uris = new ArrayList<>();
+
+        outer:
+        while (cursor.hasNext()) {
+            HashMap<String, Object> nextRow = cursor.next();
+            CrawleableUri newUri = null;
+            HashValue hashValue = null;
+            for (String key : nextRow.keySet()) {
+                if (key.equals(COLUMN_HASH_VALUE)) {
+                    String hashAsString = (String) nextRow.get(key);
+                    if (hashAsString.equals(DUMMY_HASH_VALUE)) {
+                        // no hash value in database yet for this uri
+                        continue outer;
+                    } else {
+                        hashValue = hashValueForDecoding.decodeFromString(hashAsString);
+                    }
+                } else if (key.equals(COLUMN_URI)) {
+                    try {
+                        newUri = new CrawleableUri(new URI((String) nextRow.get(key)));
+                    } catch (URISyntaxException e) {
+                        LOGGER.error("Error while constructing an uri: " + nextRow.get(key));
+                    }
+                }
+            }
+            newUri.setHashValue(hashValue);
+            uris.add(newUri);
+        }
+        cursor.close();
+        return uris;
+    }
```

Comment:
You already have a column for hash values :open_mouth: :question: 

Why are you not using it for directly searching on it the hash values of the new URIs?

Reply:
@MichaelRoeder Right now, we store the hash values in the KnownUriFilter, not in the triple store. Therefore, we have a column for the hash value here. But you're write. It would be much more efficient to ask the KnownUriFilter for all uris with same hash values. I will do it like that.